### DISCUSSION TITLE
Need help with alignment

### DISCUSSION THREAD

**Discussion by Jimmy-Z** [WerWolv/ImHex#1963] [Get Help] [answered]
I'm trying to align a struct to u32, couldn't find any example, so I wrote this:
```
fn pad32(u32 len) {
    return - len & 3;
};

struct Ptr0Str0 {
    u32 p;
    u8; // zero
    char s[];
    u8; // zero
    u8 pad[pad32(std::core::member_count(s) + 2)];
};

Ptr0Str0 array[2] @ 0;
```

then I got this:
```
E: runtime error: Cannot cast value to type 'pattern'.
```
sorry the entire error log is rather long but for whatever reason, imhex does not allow me to copy from the console.

**Comment by paxcut** [comment, marked as answer]
For alignment you can find the align to data type documented [here](https://docs.werwolv.net/pattern-language/libraries/std/mem.pat#std-mem-alignto). As for the error maybe you could use `sizeof`instead of `member_count`. Note that `char s[];` will read up to and including the zero. You should be able to use ctr-a to select the console text (you may need to click on it first) and ctrl-c to copy it to the clipboard. If that fails you can always post a screen shot

**Reply by Jimmy-Z** [reply]
Thanks, I know `char s[]` will take that trailing zero, the struct just had an extra zero there (like the zero before the string, I have no idea why).

**Comment by Jimmy-Z** [comment]
Thanks for the quick and really helpful reply.

Strangely after restarting imhex it allowed me to copy the console (I'm sure I did select the console last time, but ctrl-c doesn't work and "copy" in the context menu is grayed out, it's still grayed out this time but ctrl-c worked)

```
E: [ Stack Trace ]
E: D:\opt\imhex\includes\std/core.pat:107:56
E: 107 | d::core::member_count(pattern);  
E:                                     ^
E: <Source Code>:35:43
E: 35 | d32(std::core::member_count(s) + 2)];
E:                                    ^
E: <Source Code>:30:8
E: 30 | struct Ptr0Str0 {
E:             ^^^^^^^^
E: <Source Code>:65:5
E: 65 |     Ptr0Str0 a0l1[a0.len]
E:           ^^^^^^^^
E: <Source Code>:51:8
E: 51 | struct KeyStore {
E:             ^^^^^^^^
E: <Source Code>:118:1
E: 118 | KeyStore f @ 0;
E:       ^^^^^^^^


E: runtime error: Cannot cast value to type 'pattern'.

I: Evaluation took 0.0088974s
```
KeyStore is the struct it resides in.

After changing count_member to sizeof, it worked, I still wonder why though.

AlignTo also works, I might have stumbled upon it previously but had given up after failing to figure out how to use it.

Thanks again!

**Reply by paxcut** [reply]
Yes, currently the context menu shown if you right click on the console is the one that belongs to the pattern editor so it can't be used. That has been fixed but I haven't submitted a pr because it depends on changes that havent been reviewed yet. Ctrl-c should always work but the console needs to get focus first thats why I mentioned to left click it first.

About the reason why member_count is failing I am not really sure. I suspect there is a bug somewhere but I need to look into it first. I know arrays of built-in types are not considered patterns but that shouldn't be a restriction for member_count.

You're welcome, glad to be able to help.
